### ISSUE TITLE
Task-based MessageReceived handler

### ISSUE THREAD

**Issue by softeering** [criteo/kafka-sharp#55]
Does the ClusterClient provide a Async (i.e. Task based) handler for MessageReceived?

From the code here: https://github.com/criteo/kafka-sharp/blob/master/kafka-sharp/kafka-sharp/Public/ClusterClient.cs#L331
I could face some issues (especially in case of an exception being thrown in the handler) if I need to execute some asynchronous code there since the framework won't await on VOID (returned by the Action<...>)

**Comment by verdie-g** [comment]
Hello, you can wrap your code inside the `MessageReceived` into a `Task.Run` or enqueue the message received in a queue and dequeue it from another task. Would that work for you?

**Comment by softeering** [comment]
It "would" work but that's not exactly what I'm looking for.
I don't want to create a "new thread" but simply execute some code that is asynchronous (call to a database for example).
To me, the ClusterClient should provide an asynchronous handler in addition to the existing synchronous one to allow the user of the library to handle async operations when a message is received.

Something like
```
this._cluster.MessageReceivedAsync += async record =>
{
    await this._database.WriteAsync(record.Value...);
};
```

Where MessageReceivedAsync property would be
```
public Func<RawKafkaRecord, Task> MessageReceivedAsync { get; set; }
```

Does this make sense?

**Comment by verdie-g** [comment]
A `Task.Run` will not necessarily create a thread, the work will be pushed to the local queue of the thread and then be dequeue by the same thread it has nothing else to do. Also, if a `MessageReceivedAsync` existed, that would mean that the driver would `await` on the task you provide and postpone everything else it was doing (which can be pretty bad).

**Comment by softeering** [comment]
Yes I know, it may or may not create a new Thread but that's not the question here since calling Task.Run is not what I'm looking for.

When your client calls the MessageReceived handler, does it wait for the process completion or is it a fire-and-forget call? (I presume it waits for completion since the comment in the code says that we can use the handler to throttle the processing)

If synchronous, what does the caller do when:
- an exception is thrown by the handler
- the process completes successfully?

Thanks

**Comment by verdie-g** [comment]
`MessageReceived` is a C# event, so the handlers are executed synchronously in the same thread as the one that raised the event and handlers should not throw as the driver can not how to handle the exception.

**Comment by softeering** [comment]
That's what I understood and fair enough about "not throwing an exception" from the handler code.
Now back to the async version of the handler. Whether I defined the handler as
```
this._cluster.MessageReceived += async record => { ... };
```
or use Task.Run within the handler code block, I would break the synchronous behavior of the caller since the library won't be able to AWAIT against VOID (basically the handler won't be executed synchronously anymore).

With the current version of the MessageReceived handler, how could we call asynchronous code within the handler without breaking the synchronous behavior of the library?

**Comment by softeering** [comment]
Any update on this?
Any plan on supporting an asynchronous handler on message received?

Thanks

**Comment by verdie-g** [comment]
Hello, there is no plan on supporting that feature because it is not good to do long running operation in the callback. If you really want to do that you can synchronously wait on your Task or use [Kafka Confluent](https://github.com/confluentinc/confluent-kafka-dotnet#basic-consumer-example) that makes you write your own consuming loop.

**Comment by softeering** [comment]
Not sure why we're talking about long-running operation here.

Let's say I have a use-case where I want to consume a Kafka topic and write the record received in the handler to a database. 
With your current implementation, I'd be doing:
```csharp
this._cluster.MessageReceived += record =>
{
	this._database.ExecuteQuery("INSERT INTO TableX VALUES (?, ?, ?)", 1, "", true);
	this._logger.LogInformation("Record saved");
};
```

While what I'd like to be able to do is the following:
```csharp
this._cluster.MessageReceived += async record =>
{
	await this._database.ExecuteQueryAsync("INSERT INTO TableX VALUES (?, ?, ?)", 1, "", true);
	this._logger.LogInformation("Record saved");
};
```
And this, without changing the synchronous behavior of your library.

Operation is basically the same but would leverage non-blocking I/O

**Comment by verdie-g** [comment]
MessageReceived should only be used to enqueue the record and process it elsewhere (not in a Kafka thread). Also, I encourage you to use the Confluent Kafka driver as we are currently migrating to it.

**Comment by softeering** [comment]
ok, will move to Confluent Kafka then
thanks
